direct/tpm2/certify_test.go

```diff
@@ -160,3 +160,100 @@ func TestCertify(t *testing.T) {
 		t.Errorf("Attested buffer is different from original buffer")
 	}
 }
+
+func TestCreateAndCertifyCreation(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	public := tpm2b.Public{
+		PublicArea: tpmt.Public{
+			Type:    tpm.AlgRSA,
+			NameAlg: tpm.AlgSHA256,
+			ObjectAttributes: tpma.Object{
+				SignEncrypt:         true,
+				Restricted:          true,
+				FixedTPM:            true,
+				FixedParent:         true,
+				SensitiveDataOrigin: true,
+				UserWithAuth:        true,
+				NoDA:                true,
+			},
+			Parameters: tpmu.PublicParms{
+				RSADetail: &tpms.RSAParms{
+					Scheme: tpmt.RSAScheme{
+						Scheme: tpm.AlgRSASSA,
+						Details: tpmu.AsymScheme{
+							RSASSA: &tpms.SigSchemeRSASSA{
+								HashAlg: tpm.AlgSHA256,
+							},
+						},
+					},
+					KeyBits: 2048,
+				},
+			},
+		},
+	}
+
+	PCR7, err := CreatePCRSelection([]int{7})
+	if err != nil {
+		t.Fatalf("Failed to create PCRSelection")
+	}
+	pcrSelection := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      tpm.AlgSHA1,
+				PCRSelect: PCR7,
+			},
+		},
+	}
+
+	createPrimary := CreatePrimary{
+		PrimaryHandle: tpm.RHEndorsement,
+		InPublic:      public,
+		CreationPCR:   pcrSelection,
+	}
+	rspCP, err := createPrimary.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Failed to create primary: %v", err)
+	}
+	flushContext := FlushContext{FlushHandle: rspCP.ObjectHandle}
+	defer flushContext.Execute(thetpm)
+
+	inScheme := tpmt.SigScheme{
+		Scheme: tpm.AlgRSASSA,
+		Details: tpmu.SigScheme{
+			RSASSA: &tpms.SchemeHash{
+				HashAlg: tpm.AlgSHA256,
+			},
+		},
+	}
+
+	certifyCreation := CertifyCreation{
+		SignHandle: AuthHandle{
+			Handle: rspCP.ObjectHandle,
+			Name:   rspCP.Name,
+			Auth:   PasswordAuth([]byte{}),
+		},
+		ObjectHandle: NamedHandle{
+			Handle: rspCP.ObjectHandle,
+			Name:   rspCP.Name,
+		},
+		CreationHash:   rspCP.CreationHash,
+		InScheme:       inScheme,
+		CreationTicket: rspCP.CreationTicket,
+	}
+
+	rspCC, err := certifyCreation.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Failed to certify creation: %v", err)
+	}
+
+	attName := rspCC.CertifyInfo.AttestationData.Attested.Creation.ObjectName.Buffer
+	pubName := rspCP.Name.Buffer
+	if !cmp.Equal(attName, pubName) {
+		t.Fatalf("Attested name: %v does not match returned public key: %v.", attName, pubName)
+	}
```

Comment:
can you also validate the signature? that's an important part of CertifyCreation.

Reply:
Yup. Resolved.